core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala

```diff
@@ -387,6 +388,14 @@ private[spark] class MemoryStore(
 
   def remove(blockId: BlockId): Boolean = memoryManager.synchronized {
     val entry = entries.synchronized {
+      entries.get(blockId) match {
+        case e: DeserializedMemoryEntry[_] => e.value.foreach(v => v match {
+          case o: AutoCloseable =>
```

Comment:
OK to this would cover plain Closeable too. I get the idea. On the one hand I don't think we _know_ it's safe to close, but also can't think of a case where you wouldn't, if it is closeable. Nothing else likely has a reference to it anyway.

The downside here is that you spend time closing, and we're in a synchronized block here, which is a little worrying. But it also may encounter an exception. Do you want to try-catch for NonFatal here so that it doesn't fail if close fails?

Reply:
I suppose you can at least only synchronize on entries to remove and get the remove elements; the rest can happen after the synchronized block. We're still holding the memoryManager lock here though.

Reply:
@srowen , I fixed the codes based on your last review, 1) directly using foreach { instead of foreach(v => v {  2) I added an async call here, so I hope close functions won't hold memoryManager lock for too long 3) Added a try catch to skip NonFatal exceptions.

Reply:
In general, semantics of `AutoCloseable` has an expectation that it would be closed when released. I dont think we can make that assumption about `Closeable`.
I have not looked in greater detail if this assumption holds for Closeable for this usecase though.

Reply:
@mridulm , I see, to me, I think AutoCloseable is mostly commonly used, and close function implemented should be safe to close even more than once if it already be called by exception. Anyway, I can create a self-defined trait maybe named as "ManualCloseable", then any class extends from this "ManualCloseable" can be manual closed here, does that makes sense to you? @srowen , what do you think?

Reply:
I wouldn't make a new trait, no. I think anything AutoCloseable or Closeable is, well, closeable when done and should be closed at a point like this when we are sure (?) there are no other users of it.

Reply:
@srowen , thx, so ... Using AutoCloseable here as a hint to do manual close should be OK, am I understanding correctly?

Reply:
I think this is OK, as AutoCloseable is a supertype of Closeable, so this would handle both cases.